crates/ecs/src/systems/mod.rs

```diff
@@ -361,46 +517,96 @@ impl<Component: Debug + Send + Sync + 'static + Sized> SystemParameter for Read<
     );
 
     fn borrow<'world>(
         world: &'world World,
         archetypes: &[ArchetypeIndex],
     ) -> SystemParameterResult<Self::BorrowedData<'world>> {
-        Ok((0, {
-            let component_vecs = world
-                .borrow_component_vecs_mut::<Component>(archetypes)
-                .map_err(SystemParameterError::BorrowComponentVecs)?;
-
-            component_vecs
-                .into_iter()
-                .map(|component_vec| (0, component_vec))
-                .collect()
-        }))
-    }
-
-    unsafe fn fetch_parameter(borrowed: &mut Self::BorrowedData<'_>) -> Option<Option<Self>> {
-        let (ref mut current_archetype, archetypes) = borrowed;
-        if let Some((ref mut component_index, Some(component_vec))) =
-            archetypes.get_mut(*current_archetype)
-        {
+        let component_vecs = world
+            .borrow_component_vecs_mut::<Component>(archetypes)
+            .map_err(SystemParameterError::BorrowComponentVecs)?;
+
+        Ok(component_vecs)
+    }
+
+    fn split_borrowed_data<'borrowed>(
+        borrowed: &'borrowed mut Self::BorrowedData<'_>,
+        segment: FixedSegment,
+    ) -> Vec<Self::SegmentData<'borrowed>> {
+        let (segment_size, segment_count) = match segment {
+            FixedSegment::Single => {
+                return vec![(
+                    0,
+                    borrowed
+                        .iter_mut()
+                        .flatten()
+                        .map(|component_vec| (0, component_vec.as_mut_slice()))
+                        .collect::<Vec<_>>(),
+                )];
+            }
+            FixedSegment::Size {
+                segment_size,
+                segment_count,
+            } => (segment_size, segment_count),
+        };
+        let mut segments = Vec::with_capacity(segment_count);
+
+        let mut current_segment_size = 0;
+        let mut current_segment = vec![];
+
+        for component_vec in borrowed.iter_mut().flatten() {
+            if segment_size - current_segment_size > component_vec.len() {
+                current_segment_size += component_vec.len();
+                current_segment.push((0, component_vec.as_mut_slice()));
+                continue;
+            }
+
+            let (left, right) = component_vec.split_at_mut(segment_size - current_segment_size);
+
+            if !left.is_empty() {
+                current_segment.push((0, left));
+                segments.push((0, current_segment));
+                current_segment = vec![];
+                current_segment_size = 0;
+            }
+
+            let chunks = right.chunks_mut(segment_size);
+            for chunk in chunks {
+                if chunk.len() == segment_size {
+                    segments.push((0, vec![(0, chunk)]));
+                } else {
+                    current_segment_size += chunk.len();
+                    current_segment.push((0, chunk));
+                }
+            }
```

Comment:
> Looks like you've got two different ways of handling the remainder (one way in Read and now another one here). Is there a reason for that? Or maybe you just wrote one first and then came up with a better way of doing it later and wrote the other.

The `Write` implementation could not use `chunks_exact_mut` and `into_remainder` because both `into_iter` and `into_remaider` takes the ownership of the iterator. This is why the `Write` implementation is using `chunks_mut` instead and an if expression to check if it is the remainder.

> Either way, the implementations of split_borrowed_data are very similar, could you extract that into a helper-function and use that in both implementations?

The code looks very similar, but most of the types are different and immutable/mutable version of functions are used for the different implementation. This makes it difficult to use helper functions.

Reply:
Alright, fair enough 👍